File: python3/jump_search.py
# Jump sort will jump some steps
# When the next step is greater than finding elements
# It will get back to previous element
# And perform linear sort

# Time Complexity is O(n^0.5)


def jump_search(arr, find):
    length = len(arr)

    step = pow(length, 0.5)

    prev = 0

    while arr[int(min(step, length) - 1)] < find:
        prev = int(step)
        step += pow(length, 0.5)
        if prev >= find:
            return -1

    while arr[prev] < find:
        prev += 1

        if prev == min(step, length):
            return -1

    if arr[prev] == find:
        return prev


element_list = [0, 2, 3, 4, 5, 6, 7, 8, 9, 10]
find = 1

index = jump_search(element_list, find)
if index:
    print(f"Element found at {index + 1} position.")
else:
    print("Element is not present in given list.")

File: Js/jump_search.js
function jump_search(arr, find) {
  let length = arr.length;

  let step = Math.sqrt(length);

  let prev = 0;

  while (arr[Math.floor(Math.min(step, length) - 1)] < find) {
    prev = Math.floor(step);
    step += Math.sqrt(length);

    if (prev >= find) {
      return -1;
    }
  }

  while (arr[prev] < find) {
    prev += 1;

    if (prev == Math.min(step, length)) {
      return -1;
    }
  }

  if (arr[prev] == find) {
    return prev;
  }

  return -1;
}

let arr = new Array(10, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 33, 35, 42, 47);
let find = 22;

let position = jump_search(arr, find);
position == -1
  ? console.log("Element was not present in given list.")
  : console.log("Element is presented at " + (position + 1) + " position.");
